In [1]:
from pathlib import Path
import sys


current_directory = Path.cwd().resolve()
project_root = next(
    (
        candidate
        for candidate in (current_directory, *current_directory.parents)
        if (candidate / "src" / "path.py").is_file()
    ),
    None,
)
if project_root is None:
    raise FileNotFoundError(
        "프로젝트 루트를 찾을 수 없습니다. 프로젝트 내부에서 실행하세요."
    )
if str(project_root) not in sys.path:
    sys.path.insert(0, str(project_root))

from src.path import (  # noqa: E402
    NATIONAL_TRAVEL_SURVEY_RAW_DATA_DIR,
    NOTEBOOKS_DIR,
    PROJECT_ROOT,
)


ROOT = PROJECT_ROOT
YEARS = (2023, 2024, 2025)
SAV_FILES = {
    2023: "2023년_국민여행조사_국내여행.SAV",
    2024: "2024년_국민여행조사_국내여행.sav",
    2025: "2025년_국민여행조사_국내여행.SAV",
}

CODEBOOK_FILES = {
    2023: "2023년_국민여행조사_코드북.csv",
    2024: "2024년_국민여행조사_코드북.csv",
    2025: "2025년_국민여행조사_코드북.csv",
}

In [2]:
import pyreadstat
import pandas as pd
import pprint

# sav 파일 경로 입력
survey_raw_data_dir = NATIONAL_TRAVEL_SURVEY_RAW_DATA_DIR

# 연도별 SAV 파일 load
metadata_by_year = {}
raw_data_by_year = {}

for year in YEARS:
    sav_path = survey_raw_data_dir / f"y{year}" / SAV_FILES[year]
    df, metadata = pyreadstat.read_sav(sav_path, apply_value_formats=False)
    raw_data_by_year[year] = df
    metadata_by_year[year] = metadata
    print(f"{year}년 SAV 파일 로드 완료 [행 수: {len(df)}, 변수 수: {len(df.columns)}]")
    display(df.head(3))  # 상위 3개 행 출력

2023년 SAV 파일 로드 완료 [행 수: 52111, 변수 수: 2368]


,ID,SA1_1,SA1_2,SA1_3,SA1_4,SA1_5,SA1_6,MON_EXP_1,MON_EXP_2,MON_EXP_3,...,국내_A_기타전체여행_휴일,국내_A_기타전체여행_명절연휴,국내_A_기타숙박여행_평일,국내_A_기타숙박여행_주말,국내_A_기타숙박여행_휴일,국내_A_기타숙박여행_명절연휴,국내_A_기타당일여행_평일,국내_A_기타당일여행_주말,국내_A_기타당일여행_휴일,국내_A_기타당일여행_명절연휴
0,11010560931_124820,0.0,0.0,0.0,0.0,0.0,0.0,2.0,2.0,2.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,11010560931_124821,1.0,0.0,0.0,0.0,0.0,0.0,1.0,2.0,2.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,11010560931_124823,0.0,0.0,0.0,0.0,0.0,0.0,2.0,2.0,2.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


2024년 SAV 파일 로드 완료 [행 수: 51754, 변수 수: 2227]


,ID,SA1_1,SA1_2,SA1_3,SA1_4,SA1_5,MON_EXP_1,MON_EXP_2,MON_EXP_3,MON_EXP_4,...,국내_여행방문지_기타당일1,국내_여행방문지_기타당일2,국내_여행방문지_기타당일3,국내_여행방문지_기타당일4,국내_여행방문지_기타당일5,국내_여행방문지_기타당일6,국내_여행방문지_기타당일7,국내_여행방문지_기타당일8,국내_여행방문지_기타당일9,국내_여행방문지_기타당일10
0,11010550271_275001,0.0,0.0,0.0,0.0,0.0,2.0,2.0,2.0,2.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,11010550271_275003,1.0,0.0,0.0,0.0,0.0,1.0,2.0,2.0,2.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,11010550271_275004,1.0,0.0,0.0,0.0,0.0,1.0,2.0,2.0,2.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


2025년 SAV 파일 로드 완료 [행 수: 52185, 변수 수: 2419]


,ID,SA1_1,SA1_2,SA1_3,SA1_4,SA1_5,MON_EXP_1,MON_EXP_2,MON_EXP_3,MON_EXP_4,...,국내_여행방문지_기타당일1,국내_여행방문지_기타당일2,국내_여행방문지_기타당일3,국내_여행방문지_기타당일4,국내_여행방문지_기타당일5,국내_여행방문지_기타당일6,국내_여행방문지_기타당일7,국내_여행방문지_기타당일8,국내_여행방문지_기타당일9,국내_여행방문지_기타당일10
0,11010530051_712854,0.0,0.0,1.0,0.0,0.0,2.0,2.0,1.0,2.0,...,9.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,11010530051_712857,0.0,0.0,0.0,0.0,0.0,2.0,2.0,2.0,2.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,11010530051_712868,0.0,0.0,0.0,0.0,0.0,2.0,2.0,2.0,2.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [5]:
codebook_by_year = {}

for year in YEARS:
    meta = metadata_by_year[year]
    rows = []

    for col in meta.column_names:
        # 변수 설명
        description = meta.column_names_to_labels.get(col, "")

        # 값 라벨
        label_set = meta.variable_to_label.get(col)
        if label_set is not None:
            values = meta.value_labels.get(label_set, {})
            value_labels = "; ".join(
                f"{k}={v}" for k, v in values.items()
            )
        else:
            value_labels = ""

        # 변수 유형
        col_type = meta.variable_measure.get(col, "")

        rows.append(
            {
                "col_name": col,
                "description": description,
                "value_labels": value_labels,
                "col_type": col_type,
            }
        )
    codebook_by_year[year] = pd.DataFrame(rows)

# 각 연도별 코드북 확인
for year, codebook in codebook_by_year.items():
    print(f"\n{year}년 코드북:")
    display(codebook.head())


2023년 코드북:


,col_name,description,value_labels,col_type
0,ID,아이디,,nominal
1,SA1_1,공통] A_여행 횟수-1. 국내 관광/휴양 여행,,scale
2,SA1_2,공통] A_여행 횟수-2. 국내 가족/친지/친구 방문 여행 - 관광/휴양 활동 포함,,scale
3,SA1_3,공통] A_여행 횟수-3. 국내 단순 가족/친지/친구 방문 - 관광/휴양 활동이 ...,,scale
4,SA1_4,공통] A_여행 횟수-4. 국내 출장/업무 여행 - 관광/휴양 활동 포함,,scale



2024년 코드북:


,col_name,description,value_labels,col_type
0,ID,아이디,,nominal
1,SA1_1,공통] A_여행 횟수-1. 국내 관광/휴양 여행,,nominal
2,SA1_2,공통] A_여행 횟수-2. 국내 가족/친지/친구 방문 여행 - 관광/휴양 활동 포함,,nominal
3,SA1_3,공통] A_여행 횟수-3. 국내 단순 가족/친지/친구 방문 - 관광/휴양 활동이 ...,,nominal
4,SA1_4,공통] A_여행 횟수-4. 국내 출장/업무 여행 - 관광/휴양 활동 포함,,nominal



2025년 코드북:


,col_name,description,value_labels,col_type
0,ID,아이디,,nominal
1,SA1_1,공통] A_여행 횟수-1. 국내 관광/휴양 여행,,nominal
2,SA1_2,공통] A_여행 횟수-2. 국내 가족/친지/친구 방문 여행 - 관광/휴양 활동 포함,,nominal
3,SA1_3,공통] A_여행 횟수-3. 국내 단순 가족/친지/친구 방문 - 관광/휴양 활동이 ...,,nominal
4,SA1_4,공통] A_여행 횟수-4. 국내 출장/업무 여행 - 관광/휴양 활동 포함,,nominal


In [6]:
# 설문조사 원변수만 추출
import re
survey_codebook_by_year = {}

for year, codebook in codebook_by_year.items():
    target_columns = [col for col in codebook["col_name"] if re.match(r"^[A-Za-z]", col)]
    survey_codebook = codebook[codebook["col_name"].isin(target_columns)]
    survey_codebook_by_year[year] = survey_codebook
    print(f"\n{year}년 설문조사 원변수:")
    display(survey_codebook)
    survey_codebook.to_csv(
        NATIONAL_TRAVEL_SURVEY_RAW_DATA_DIR / f"y{year}" / CODEBOOK_FILES[year],
        index=False,
        encoding="utf-8-sig"
    )


2023년 설문조사 원변수:


,col_name,description,value_labels,col_type
0,ID,아이디,,nominal
1,SA1_1,공통] A_여행 횟수-1. 국내 관광/휴양 여행,,scale
2,SA1_2,공통] A_여행 횟수-2. 국내 가족/친지/친구 방문 여행 - 관광/휴양 활동 포함,,scale
3,SA1_3,공통] A_여행 횟수-3. 국내 단순 가족/친지/친구 방문 - 관광/휴양 활동이 ...,,scale
4,SA1_4,공통] A_여행 횟수-4. 국내 출장/업무 여행 - 관광/휴양 활동 포함,,scale
...,...,...,...,...
1371,BMAR,공통] 혼인상태,1.0= 미혼; 2.0= 배우자 있음; 3.0= 사별; 4.0= 이혼; 5....,nominal
1372,BFAM,공통] 동거가구원수,1.0= 1인; 2.0= 2인; 3.0= 3인 이상,nominal
1373,BMON,공통] 월별,1.0= 1월; 2.0= 2월; 3.0= 3월; 4.0= 4월; 5.0= ...,nominal
1374,BARA,공통] 거주지역,1.0= 서울; 2.0= 부산; 3.0= 대구; 4.0= 인천; 5.0= ...,nominal



2024년 설문조사 원변수:


,col_name,description,value_labels,col_type
0,ID,아이디,,nominal
1,SA1_1,공통] A_여행 횟수-1. 국내 관광/휴양 여행,,nominal
2,SA1_2,공통] A_여행 횟수-2. 국내 가족/친지/친구 방문 여행 - 관광/휴양 활동 포함,,nominal
3,SA1_3,공통] A_여행 횟수-3. 국내 단순 가족/친지/친구 방문 - 관광/휴양 활동이 ...,,nominal
4,SA1_4,공통] A_여행 횟수-4. 국내 출장/업무 여행 - 관광/휴양 활동 포함,,nominal
...,...,...,...,...
1230,BMAR,공통] 혼인상태,1.0= 미혼; 2.0= 배우자 있음; 3.0= 사별; 4.0= 이혼; 5....,nominal
1231,BFAM,공통] 동거가구원수,1.0= 1인; 2.0= 2인; 3.0= 3인 이상,nominal
1232,BMON,공통] 월별,1.0= 1월; 2.0= 2월; 3.0= 3월; 4.0= 4월; 5.0= ...,nominal
1233,BARA,공통] 거주지역,1.0= 서울; 2.0= 부산; 3.0= 대구; 4.0= 인천; 5.0= ...,nominal



2025년 설문조사 원변수:


,col_name,description,value_labels,col_type
0,ID,아이디,,nominal
1,SA1_1,공통] A_여행 횟수-1. 국내 관광/휴양 여행,,nominal
2,SA1_2,공통] A_여행 횟수-2. 국내 가족/친지/친구 방문 여행 - 관광/휴양 활동 포함,,nominal
3,SA1_3,공통] A_여행 횟수-3. 국내 단순 가족/친지/친구 방문 - 관광/휴양 활동이 ...,,nominal
4,SA1_4,공통] A_여행 횟수-4. 국내 출장/업무 여행 - 관광/휴양 활동 포함,,nominal
...,...,...,...,...
1422,BMAR,공통] 혼인상태,1.0= 미혼; 2.0= 배우자 있음; 3.0= 사별; 4.0= 이혼; 5....,nominal
1423,BFAM,공통] 동거가구원수,1.0= 1인; 2.0= 2인; 3.0= 3인 이상,nominal
1424,BMON,공통] 월별,1.0= 1월; 2.0= 2월; 3.0= 3월; 4.0= 4월; 5.0= ...,nominal
1425,BARA,공통] 거주지역,1.0= 서울; 2.0= 부산; 3.0= 대구; 4.0= 인천; 5.0= ...,nominal


## 세 연도 원변수 결합을 위한 변수·코드 비교

연도별 공식 코드북의 `3. 국내여행 코드 > 1) 원변수`를 기준으로 비교한다. 세 코드북의 원변수명은 SAV에서 영문자로 시작하는 변수와 정확히 일치하므로 `^[A-Za-z]` 조건을 사용한다. `국내_월별여행경험_*`처럼 한글로 시작하는 분석변수는 제외한다. 코드는 SAV 값 라벨에 정의된 원코드를 뜻하며, 값 라벨이 없는 변수는 메타데이터만으로 코드 동일성을 판정하지 않는다.

참고 문서: `docs/report/국민여행조사/2023년_국민여행조사_코드북.pdf`, `2024년_국민여행조사_코드북.pdf`, `2025년_국민여행조사_코드북.pdf`.

In [7]:
sav_paths = sorted(
    path
    for path in survey_raw_data_dir.iterdir()
    if path.suffix.lower() == ".sav"
)

metadata_by_year = {}
for sav_path in sav_paths:
    year = sav_path.name[:4]
    _, sav_metadata = pyreadstat.read_sav(sav_path, metadataonly=True)
    metadata_by_year[year] = sav_metadata

official_original_variable_counts = {
    "2023": 1376,
    "2024": 1235,
    "2025": 1427,
}
original_variables_by_year = {
    year: [
        variable
        for variable in metadata.column_names
        if re.match(r"^[A-Za-z]", variable)
    ]
    for year, metadata in metadata_by_year.items()
}

file_summary = pd.DataFrame(
    [
        {
            "연도": year,
            "파일명": sav_path.name,
            "변수 수": len(metadata_by_year[year].column_names),
            "공식 원변수 수": len(original_variables_by_year[year]),
            "분석·기타변수 수": (
                len(metadata_by_year[year].column_names)
                - len(original_variables_by_year[year])
            ),
            "값 라벨이 있는 변수 수": len(
                metadata_by_year[year].variable_value_labels
            ),
        }
        for sav_path in sav_paths
        for year in [sav_path.name[:4]]
    ]
).sort_values("연도", ignore_index=True)

assert {
    year: len(variables)
    for year, variables in original_variables_by_year.items()
} == official_original_variable_counts

file_summary

,연도,파일명,변수 수,공식 원변수 수,분석·기타변수 수,값 라벨이 있는 변수 수
0,2023,2023년_국민여행조사_국내여행.SAV,2368,1376,992,601
1,2024,2024년_국민여행조사_국내여행.sav,2227,1235,992,595
2,2025,2025년_국민여행조사_국내여행.SAV,2419,1427,992,654


In [8]:
years = sorted(metadata_by_year)
variable_sets = {
    year: set(original_variables_by_year[year]) for year in years
}
all_variables = sorted(set.union(*variable_sets.values()))
variable_presence = pd.DataFrame(
    {
        year: [variable in variable_sets[year] for variable in all_variables]
        for year in years
    },
    index=all_variables,
).rename_axis("변수명")
variable_presence["존재 연도"] = variable_presence.apply(
    lambda row: ", ".join(year for year in years if row[year]),
    axis=1,
)

variable_presence_summary = (
    variable_presence.groupby("존재 연도", sort=False)
    .size()
    .rename("변수 수")
    .reset_index()
    .sort_values("존재 연도", ignore_index=True)
)
variable_presence_differences = variable_presence.loc[
    variable_presence["존재 연도"] != ", ".join(years)
]

display(variable_presence_summary)
display(variable_presence_differences)

,존재 연도,변수 수
0,2023,292
1,"2023, 2024, 2025",918
2,"2023, 2025",166
3,"2024, 2025",317
4,2025,26


,2023,2024,2025,존재 연도
변수명,,,,
A1,False,True,True,"2024, 2025"
A10A_1,True,False,False,2023
A10A_10,True,False,False,2023
A10A_11,True,False,False,2023
A10A_12,True,False,False,2023
...,...,...,...,...
TA8SF_4,False,False,True,2025
TB9_11,True,False,False,2023
ZQ1_1,False,True,True,"2024, 2025"


In [9]:
common_variables = sorted(set.intersection(*variable_sets.values()))
code_comparison_rows = []
for variable in common_variables:
    value_mappings = {
        year: metadata_by_year[year].variable_value_labels.get(variable, {})
        for year in years
    }
    code_sets = {year: set(value_mappings[year]) for year in years}
    defined_year_count = sum(bool(mapping) for mapping in value_mappings.values())
    code_sets_same = len(
        {frozenset(code_sets[year]) for year in years}
    ) == 1
    label_mappings_same = len(
        {repr(value_mappings[year]) for year in years}
    ) == 1

    if defined_year_count == 0:
        code_status = "값 라벨 코드 정의 없음"
    elif defined_year_count < len(years):
        code_status = "연도별 코드 정의 유무 다름"
    elif code_sets_same:
        code_status = "코드 동일"
    else:
        code_status = "코드 다름"

    row = {
        "변수명": variable,
        "코드 비교": code_status,
        "라벨 문구까지 동일": label_mappings_same,
    }
    for year in years:
        row[f"{year} 변수 설명"] = metadata_by_year[
            year
        ].column_names_to_labels.get(variable, "")
        row[f"{year} 코드"] = sorted(
            code_sets[year],
            key=lambda code: (type(code).__name__, code),
        )
        row[f"{year} 코드 라벨"] = value_mappings[year]
    code_comparison_rows.append(row)

common_code_comparison = pd.DataFrame(code_comparison_rows)
code_compatibility_summary = (
    common_code_comparison["코드 비교"]
    .value_counts(sort=False)
    .rename_axis("구분")
    .rename("변수 수")
    .reset_index()
)
code_differences = common_code_comparison.loc[
    common_code_comparison["코드 비교"] == "코드 다름"
]
code_definition_differences = common_code_comparison.loc[
    common_code_comparison["코드 비교"]
    == "연도별 코드 정의 유무 다름"
]
storage_type_difference_count = sum(
    len(
        {
            metadata_by_year[year].readstat_variable_types.get(variable)
            for year in years
        }
    )
    > 1
    for variable in common_variables
)

display(code_compatibility_summary)
print(f"저장 자료형이 다른 공통 변수: {storage_type_difference_count:,}개")
display(code_differences)
display(code_definition_differences)

,구분,변수 수
0,코드 동일,291
1,코드 다름,14
2,연도별 코드 정의 유무 다름,4
3,값 라벨 코드 정의 없음,609


저장 자료형이 다른 공통 변수: 0개


,변수명,코드 비교,라벨 문구까지 동일,2023 변수 설명,2023 코드,2023 코드 라벨,2024 변수 설명,2024 코드,2024 코드 라벨,2025 변수 설명,2025 코드,2025 코드 라벨
2,A3_1,코드 다름,False,A.[국내] Q3. 사전예약사항_1,"[1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 10.0, 11.0]","{1.0: '숙박시설(호텔, 콘도미니엄/리조트, 펜션, 유스호스텔 등)', 2.0:...",A.[국내] Q3. 여행지 선택 이유_1순위,"[1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, ...","{1.0: '여행지 지명도', 2.0: '볼거리 제공', 3.0: '저렴한 여행경비...",A.[국내]Q3. 다녀오신 여행의 여행 지역을 선택한 이유는 무엇입니까?-1순위,"[1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, ...","{1.0: '여행지 지명도', 2.0: '볼거리 제공', 3.0: '저렴한 여행경비..."
3,A3_2,코드 다름,False,A.[국내] Q3. 사전예약사항_2,"[1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 10.0, 11.0]","{1.0: '숙박시설(호텔, 콘도미니엄/리조트, 펜션, 유스호스텔 등)', 2.0:...",A.[국내] Q3. 여행지 선택 이유_2순위,"[1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, ...","{1.0: '여행지 지명도', 2.0: '볼거리 제공', 3.0: '저렴한 여행경비...",A.[국내]Q3. 다녀오신 여행의 여행 지역을 선택한 이유는 무엇입니까?-2순위,"[1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, ...","{1.0: '여행지 지명도', 2.0: '볼거리 제공', 3.0: '저렴한 여행경비..."
4,A3_3,코드 다름,False,A.[국내] Q3. 사전예약사항_3,"[1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 10.0, 11.0]","{1.0: '숙박시설(호텔, 콘도미니엄/리조트, 펜션, 유스호스텔 등)', 2.0:...",A.[국내] Q3. 여행지 선택 이유_3순위,"[1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, ...","{1.0: '여행지 지명도', 2.0: '볼거리 제공', 3.0: '저렴한 여행경비...",A.[국내]Q3. 다녀오신 여행의 여행 지역을 선택한 이유는 무엇입니까?-3순위,"[1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, ...","{1.0: '여행지 지명도', 2.0: '볼거리 제공', 3.0: '저렴한 여행경비..."
5,A4_1,코드 다름,False,A.[국내] Q4. 여행지 선택 이유_1순위,"[1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, ...","{1.0: '여행지 지명도', 2.0: '볼거리 제공', 3.0: '저렴한 여행경비...",A.[국내] Q4. 여행지에서의 활동_1,"[1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, ...","{1.0: '자연 및 풍경감상', 2.0: '음식관광(지역 맛집 등)', 3.0: ...",A.[국내]Q4. 다녀오신 여행에서 참여하신 활동은 무엇입니까?-1,"[1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, ...","{1.0: '자연 및 풍경감상', 2.0: '음식관광(지역 맛집 등)', 3.0: ..."
6,A4_2,코드 다름,False,A.[국내] Q4. 여행지 선택 이유_2순위,"[1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, ...","{1.0: '여행지 지명도', 2.0: '볼거리 제공', 3.0: '저렴한 여행경비...",A.[국내] Q4. 여행지에서의 활동_2,"[1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, ...","{1.0: '자연 및 풍경감상', 2.0: '음식관광(지역 맛집 등)', 3.0: ...",A.[국내]Q4. 다녀오신 여행에서 참여하신 활동은 무엇입니까?-2,"[1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, ...","{1.0: '자연 및 풍경감상', 2.0: '음식관광(지역 맛집 등)', 3.0: ..."
7,A4_3,코드 다름,False,A.[국내] Q4. 여행지 선택 이유_3순위,"[1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, ...","{1.0: '여행지 지명도', 2.0: '볼거리 제공', 3.0: '저렴한 여행경비...",A.[국내] Q4. 여행지에서의 활동_3,"[1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, ...","{1.0: '자연 및 풍경감상', 2.0: '음식관광(지역 맛집 등)', 3.0: ...",A.[국내]Q4. 다녀오신 여행에서 참여하신 활동은 무엇입니까?-3,"[1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, ...","{1.0: '자연 및 풍경감상', 2.0: '음식관광(지역 맛집 등)', 3.0: ..."
8,A5_1,코드 다름,False,A.[국내] Q5. 여행지에서의 활동_1,"[1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, ...","{1.0: '자연 및 풍경감상', 2.0: '음식관광(지역 맛집 등)', 3.0: ...",A.[국내] Q5. 여행 정보 획득 경로_1순위,"[1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0]","{1.0: '인터넷 사이트/모바일 앱(PC/스마트폰)', 2.0: '기사 및 방송 ...","A.[국내]Q5. 다녀오신 여행을 위한 여행 정보(먹거리, 즐길거리, 볼거리, #숙...","[1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0]","{1.0: '인터넷 사이트/모바일 앱(PC/스마트폰)', 2.0: '기사 및 방송 ..."
9,A5_2,코드 다름,False,A.[국내] Q5. 여행지에서의 활동_2,"[1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, ...","{1.0: '자연 및 풍경감상', 2.0: '음식관광(지역 맛집 등)', 3.0: ...",A.[국내] Q5. 여행 정보 획득 경로_2순위,"[1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0]","{1.0: '인터넷 사이트/모바일 앱(PC/스마트폰)', 2.0: '기사 및 방송 ...","A.[국내]Q5. 다녀오신 여행을 위한 여행 정보(먹거리, 즐길거리, 볼거리, #숙...","[1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0]","{1.0: '인터넷 사이트/모바일 앱(PC/스마트폰)', 2.0: '기사 및 방송 ..."
10,A5_3,코드 다름,False,A.[국내] Q5. 여행지에서의 활동_3,"[1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, ...","{1.0: '자연 및 풍경감상', 2.0: '음식관광(지역 맛집 등)', 3.0: ...",A.[국내] Q5. 여행 정보 획득 경로_3순위,"[1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0]","{1.0: '인터넷 사이트/모바일 앱(PC/스마트폰)', 2.0: '기사 및 방송 ...","A.[국내]Q5. 다녀오신 여행을 위한 여행 정보(먹거리, 즐길거리, 볼거리, #숙...","[1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0]","{1.0: '인터넷 사이트/모바일 앱(PC/스마트폰)', 2.0: '기사 및 방송 ..."
209,D_TRA1_CASE,코드 다름,False,국내] B_여행1차_여행유형,"[1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0]","{1.0: '국내 관

,변수명,코드 비교,라벨 문구까지 동일,2023 변수 설명,2023 코드,2023 코드 라벨,2024 변수 설명,2024 코드,2024 코드 라벨,2025 변수 설명,2025 코드,2025 코드 라벨
11,A9,연도별 코드 정의 유무 다름,False,A.[국내] Q9. 총합-지출 비용[단위:원],[],{},A.[국내] Q9. 전반적 만족도,"[1.0, 2.0, 3.0, 4.0, 5.0]","{1.0: '매우 불만족', 2.0: '불만족', 3.0: '보통', 4.0: '만...",A.[국내]Q9. 다녀오신 여행지에 대해 전반적으로 얼마나 만족하셨습니까?,"[1.0, 2.0, 3.0, 4.0, 5.0]","{1.0: '매우 불만족', 2.0: '불만족', 3.0: '보통', 4.0: '만..."
12,A9A_1,연도별 코드 정의 유무 다름,False,"A.[국내] Q9. 1-1. 여행사(온라인, 오프라인) 구입-지출 비용[단위:원]",[],{},A.[국내] Q9-1. 항목별 만족도-01)자연경관,"[1.0, 2.0, 3.0, 4.0, 5.0, 9.0]","{1.0: '매우 불만족', 2.0: '불만족', 3.0: '보통', 4.0: '만...",A.[국내]Q9-1. 다녀오신 여행지의 다음의 각 항목에 대해 얼마나 만족하셨습니까...,"[1.0, 2.0, 3.0, 4.0, 5.0, 9.0]","{1.0: '매우 불만족', 2.0: '불만족', 3.0: '보통', 4.0: '만..."
13,A9A_2,연도별 코드 정의 유무 다름,False,"A.[국내] Q9. 1-2. 여행사 이외(홈쇼핑, SNS에서 개인이 하는 업체 등)...",[],{},A.[국내] Q9-1. 항목별 만족도-02)문화유산,"[1.0, 2.0, 3.0, 4.0, 5.0, 9.0]","{1.0: '매우 불만족', 2.0: '불만족', 3.0: '보통', 4.0: '만...",A.[국내]Q9-1. 다녀오신 여행지의 다음의 각 항목에 대해 얼마나 만족하셨습니까...,"[1.0, 2.0, 3.0, 4.0, 5.0, 9.0]","{1.0: '매우 불만족', 2.0: '불만족', 3.0: '보통', 4.0: '만..."
14,A9A_3,연도별 코드 정의 유무 다름,False,A.[국내] Q9. 1-3. 관광 가이드 활동비(팁)-지출 비용[단위:원],[],{},A.[국내] Q9-1. 항목별 만족도-03)교통,"[1.0, 2.0, 3.0, 4.0, 5.0, 9.0]","{1.0: '매우 불만족', 2.0: '불만족', 3.0: '보통', 4.0: '만...",A.[국내]Q9-1. 다녀오신 여행지의 다음의 각 항목에 대해 얼마나 만족하셨습니까...,"[1.0, 2.0, 3.0, 4.0, 5.0, 9.0]","{1.0: '매우 불만족', 2.0: '불만족', 3.0: '보통', 4.0: '만..."


### 공식 코드북에서 확인한 2023년 A문항 이동

2023년과 2024·2025년은 A영역의 문항 번호가 이동했다. 같은 변수명만 기준으로 합치면 서로 다른 문항이 한 열에 들어가므로, 아래처럼 문항 의미를 기준으로 2023년 변수명을 명시적으로 매핑해야 한다.

| 문항 의미 | 2023년 | 2024·2025년 기준명 | 주의사항 |
|---|---|---|---|
| 여행사 상품 구매 | `A2`, `A2A` | `A1`, `A1A` | 2023년은 비구매 코드 정의가 다르므로 분기 확인 |
| 개별 사전예약 | `A3_*` | `A2_*` | 2024년부터 예약 항목 8·9 추가 |
| 여행지 선택 이유 | `A4_*` | `A3_*` | 공통 코드 1~15 사용 |
| 여행 활동 | `A5_*`, `A5A` | `A4_*`, `A4A` | 코드 16의 라벨 범위가 2025년에 확대 |
| 정보 획득 경로 | `A6_*`, `A6A_*` | `A5_*`, `A5A_*` | 인터넷 경로 항목이 2024년부터 확대 |
| 동반자 | `A7`, `A7A`, `A7B_*` | `A6`, `A6A`, `A6B_*` | 동반자 유형이 2024년부터 세분화 |
| 출발 전 지출 | `A8_*` | `A7_*` | 같은 지출 항목끼리 매핑 |
| 여행지·이동 지출 | `A9*` | `A8*` | 접미부를 포함한 명시적 대응표 필요 |
| 만족도 | `A10`, `A10A_*` | `A9`, `A9A_*` | 같은 이름의 2023년 `A9*`와 합치면 안 됨 |
| 재방문 의향 | `A11` | `A10` | 5점 척도 |
| 추천 의향 | `A12` | `A11` | 5점 척도 |

2023년 `A1_*` 주요 이동수단은 2024년 이후 같은 형태의 직접 대응 변수가 없고, 2024년 이후의 `A6C` 반려동물 동반 여부와 `A12_*` 여행 효과 등은 신규 문항이다. 이런 변수는 삭제하지 않고 연도 한정 변수로 보존한다.

In [10]:
trip_case_columns = [f"D_TRA{slot}_CASE" for slot in range(1, 6)]
observed_case_rows = []
for sav_path in sav_paths:
    year = sav_path.name[:4]
    case_data, _ = pyreadstat.read_sav(
        sav_path,
        usecols=trip_case_columns,
        apply_value_formats=False,
    )
    for variable in trip_case_columns:
        observed_case_rows.append(
            {
                "연도": year,
                "변수명": variable,
                "실제 관측 코드": sorted(
                    case_data[variable].dropna().unique().tolist()
                ),
            }
        )

observed_trip_case_codes = pd.DataFrame(observed_case_rows)
observed_trip_case_codes

,연도,변수명,실제 관측 코드
0,2023,D_TRA1_CASE,"[1.0, 2.0, 3.0, 4.0, 5.0]"
1,2023,D_TRA2_CASE,"[1.0, 2.0, 3.0, 4.0, 5.0]"
2,2023,D_TRA3_CASE,"[1.0, 2.0, 3.0, 4.0, 5.0]"
3,2023,D_TRA4_CASE,"[1.0, 2.0, 3.0, 4.0, 5.0]"
4,2023,D_TRA5_CASE,"[2.0, 3.0, 5.0]"
5,2024,D_TRA1_CASE,"[1.0, 2.0, 3.0, 4.0, 5.0]"
6,2024,D_TRA2_CASE,"[1.0, 2.0, 3.0, 4.0, 5.0]"
7,2024,D_TRA3_CASE,"[1.0, 2.0, 3.0, 5.0]"
8,2024,D_TRA4_CASE,"[1.0, 3.0, 4.0, 5.0]"
9,2024,D_TRA5_CASE,"[1.0, 3.0]"


### 원변수를 최대한 보존하는 결합 방법

1. 원본 SAV는 수정하지 않고 연도별 원변수만 읽는다. 결합 키는 `YEAR + ID`로 만들고, 원래 변수명과 원코드를 보존한다. `ID`가 연도 사이에도 고유하다고 가정하지 않는다.
2. 원변수 합집합 1,719개를 기준으로 outer concat하면 모든 변수를 보존할 수 있다. 세 연도 공통 변수는 918개이며, 나머지는 없는 연도를 구조적 결측으로 두고 `AVAILABLE_연도` 정보를 별도 관리한다.
3. 2023년 A문항은 위 의미 대응표에 따라 2024·2025년 기준의 별도 `canonical_name`을 만든다. 원변수 열을 덮어쓰지 말고 분석용 표에서만 canonical 열을 생성한다.
4. 같은 문항이라도 새 범주가 추가된 경우 기존 범주를 삭제하지 않는다. 공통 범주는 그대로 유지하고, 신규 범주는 해당 연도에서만 나타나는 유효 범주로 둔다. 범주를 합쳐야 할 때만 별도의 상위 범주 변수를 만든다.
5. `D_TRA*_CASE`는 2023년 메타데이터에 6~9가 추가 정의되어 있지만 실제 국내여행 SAV의 관측값은 세 연도 모두 1~5 범위다. 원코드 1~5를 유지해 결합할 수 있다.
6. `WT_DOM`은 각 연도 안의 모집단 추정에 사용한다. 세 연도 가중치를 단순 합산해 고유 인구로 해석하지 않고, 연도별 지표를 먼저 산출한 뒤 비교하거나 분석 목적에 맞는 pooled 추정량을 별도로 정의한다.
7. 최종 분석표에는 `YEAR`, `ID`, `source_name`, `canonical_name`, `original_code`, `harmonized_code`, `availability`, `structural_missing_reason`을 추적할 수 있는 변수 대응표를 함께 둔다.

현재 동일 변수명 기준 코드표에서 `코드 다름`으로 나온 14개는 A문항 번호 이동 9개와 `D_TRA1_CASE`~`D_TRA5_CASE` 5개다. `연도별 코드 정의 유무 다름` 4개는 2023년 지출 문항 `A9`, `A9A_1`~`A9A_3`과 2024·2025년 만족도 문항이 같은 이름을 사용한 경우다. 따라서 이 18개를 단순 재코딩하기보다 먼저 의미 기준으로 열을 분리·이름 매핑해야 한다.